### ISSUE TITLE
`stdenv.hostPlatform != stdenv.buildPlatform` ----> `stdenv.isCrossCompiling`

### ISSUE THREAD

**Issue by kvtb** [NixOS/nixpkgs#130349]
It think, there should be `stdenv.isCrossCompiling` to replace numerous `stdenv.hostPlatform != stdenv.buildPlatform`, `stdenv.hostPlatform.system != stdenv.buildPlatform.system`, `stdenv.hostPlatform.config != stdenv.buildPlatform.config` scattered across nixpkgs.

Besides the confusion of too  many forms to express the idea of cross-compilation, all of them are incorrect, because there is an exception of `i686` compiled on `x86_64`, which is obviously not cross-compilation

So it should be something like
```nix
isCrossCompiling = stdenv.buildPlatform.system != stdenv.hostPlatform.system &&
  !(stdenv.buildPlatform.isx86_64 && stdenv.hostPlatform.i686);
```

@matthewbauer

**Comment by Mic92** [comment]
I think this comparison was introduced by @Ericson2314 intentionally

**Comment by Mindavi** [comment]
Technically it (x86-64 to i686) is cross-compilation, but indeed it might be handy to handle that differently, because the binaries can run on 64-bit platforms.

Otherwise, this might be nice to have to make intentions more explicit.

**Comment by Ericson2314** [comment]
I don't want this. I think people have many preconceived notions of cross compiling, it is better to be explicit. For example, for compilers we also do `target != host` checks, and then people will want a short-hand for that too, etc. etc.

For the second part, we have `isCompatible` https://github.com/NixOS/nixpkgs/blob/1c586291425a2e2c5683db88a829dbc419f23e45/lib/systems/parse.nix#L132-L215 to take those situations into account. But again, that is *not always* the notion one wants. For example, you can run 32-bit code on many 64-bit arches, but you cannot mix SOs between the two.

An `isCrossCompiling` that does or doesn't do the `isCompatible` thing will also create confusion in the case when it's choice is no appropriate.

Basically, `foo != bar` is short enough that clarity >>> brevity.
